### PR TITLE
Fix GiveUnitCrateAction and SpawnMPUnits being limited to MobileInfo

### PR DESCRIPTION
**Pull request by abcdefg30**
You can now also spawn aircraft and crates.

### CODE REVIEW

**Comment by reaperrr** [comment]
This should also allow aircraft as starting units. Nice!

**Comment by abcdefg30** [comment]
Good idea! Added it.

File: OpenRA.Mods.Common/Traits/World/SpawnMPUnits.cs [hunk 1]
```diff
@@ -86,16 +86,18 @@ void SpawnUnitsForPlayer(World w, Player p, CPos sp)
...
+				var validCells = supportSpawnCells.Where(c => ip.CanEnterCell(w, null, c));
 				if (!validCells.Any())
 				{
 					Log.Write("debug", "No cells available to spawn starting unit {0} for player {1}".F(s, p));
 					continue;
 				}
 
 				var cell = validCells.Random(w.SharedRandom);
-				var subCell = mi.SharesCell ? w.ActorMap.FreeSubCell(cell) : 0;
+				var ios = actorRules.TraitInfo<IOccupySpaceInfo>();
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/SpawnMPUnits.cs]: IPositionable inherits IOccupySpace, so this should not be required.

Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/World/SpawnMPUnits.cs]: We use the `Info` classes and `IPositionableInfo` doesn't inherit `IOccupiesSpaceInfo` afaik.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/SpawnMPUnits.cs]: That is a bug that needs to be fixed.

Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/World/SpawnMPUnits.cs]: Fixed.

**Review by GraionDilach — APPROVED** [review]
:+1:

**Review by GraionDilach — CHANGES_REQUESTED** [review]
Wait... no, you should drop IOccupySpaceInfo from CrateInfo, AircraftInfo and co. now.

**Comment by abcdefg30** [comment]
Fixed.

**Review by GraionDilach — APPROVED** [review]
:+1:

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 2]
```diff
@@ -94,6 +94,25 @@ public class AircraftInfo : ITraitInfo, IPositionableInfo, IFacingInfo, IOccupyS
 
 		public IReadOnlyDictionary<CPos, SubCell> OccupiedCells(ActorInfo info, CPos location, SubCell subCell = SubCell.Any) { return new ReadOnlyDictionary<CPos, SubCell>(); }
 		bool IOccupySpaceInfo.SharesCell { get { return false; } }
+
+		// Used to determe if an aircraft can spawn landed
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: > determe

File: OpenRA.Mods.Common/Traits/Mobile.cs [hunk 3]
```diff
@@ -380,7 +381,7 @@ public int MovementCostToEnterCell(WorldMovementInfo worldMovementInfo, Actor se
 	}
 
 	public class Mobile : ConditionalTrait<MobileInfo>, INotifyCreated, IIssueOrder, IResolveOrder, IOrderVoice, IPositionable, IMove,
-		IFacing, ISync, IDeathActorInitModifier, INotifyAddedToWorld, INotifyRemovedFromWorld, INotifyBlockingMove, IActorPreviewInitModifier
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Mobile.cs]: erm, why are you removing `ISync`?

Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Mobile.cs]: `ConditionalTrait<MobileInfo>` implements `ISync` already.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs (hunk 2 above)
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Woops, fixed.

**Comment by reaperrr** [comment]
Spawning `heli` either as starting unit or crate action in RA doesn't work for me. There isn't any crash, but the heli just doesn't appear, either.

**Comment by abcdefg30** [comment]
> Spawning `heli` either as starting unit or crate action in RA doesn't work for me. There isn't any crash, but the heli just doesn't appear, either.

You're probably missing the correct `LandableTerrainTypes` definitions in yaml.

**Comment by reaperrr** [comment]
> You're probably missing the correct LandableTerrainTypes definitions in yaml.

Ok, that works.
In the future I'd like to see support for aircraft starting mid-air, so the terrain doesn't matter, but for this PR it's good enough.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/79f8f3f9af0d1e1b1314d5c386bdd5c3396794c7...5cc60976fd73e0c093a4615ff80f93858edcbedb)

**Pull request merged**